plugins/intel-gsc/fu-igsc-device.c

```diff
@@ -0,0 +1,844 @@
+{
+	struct gsc_fwu_heci_version_req req = {.header.command_id =
+						   GSC_FWU_HECI_COMMAND_ID_GET_IP_VERSION,
+					       .partition = partition};
+	guint8 res_buf[100];
+	struct gsc_fwu_heci_version_resp *res = (struct gsc_fwu_heci_version_resp *)res_buf;
+
+	if (!fu_igsc_device_command(self,
+				    (const guint8 *)&req,
+				    sizeof(req),
+				    res_buf,
+				    sizeof(struct gsc_fwu_heci_version_resp) + bufsz,
+				    error)) {
+		g_prefix_error(error, "invalid HECI message response: ");
+		return FALSE;
+	}
+	if (!fu_igsc_device_heci_validate_response_header(self,
+							  &res->response,
+							  req.header.command_id,
+							  error))
+		return FALSE;
+	if (res->partition != partition) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid HECI message response payload: 0x%x: ",
+			    res->partition);
+		return FALSE;
+	}
+	if (bufsz > 0 && res->version_length != bufsz) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid HECI message response version_length: 0x%x, expected 0x%x: ",
+			    res->version_length,
+			    (guint)bufsz);
+		return FALSE;
+	}
+	if (buf != NULL) {
+		if (!fu_memcpy_safe(buf,
+				    bufsz,
+				    0x0, /* dst */
+				    res->version,
+				    res->version_length,
+				    0x0, /* src*/
+				    res->version_length,
+				    error)) {
+			return FALSE;
+		}
+	}
+
+	/* success */
+	return TRUE;
+}
+
+gboolean
+fu_igsc_device_get_aux_version(FuIgscDevice *self,
+			       guint32 *oem_version,
+			       guint16 *major_version,
+			       guint16 *major_vcn,
+			       GError **error)
+{
+	struct gsc_fw_data_heci_version_req req = {
+	    .header.command_id = GSC_FWU_HECI_COMMAND_ID_GET_GFX_DATA_UPDATE_INFO};
+	struct gsc_fw_data_heci_version_resp res = {0x0};
+
+	if (!fu_igsc_device_command(self,
+				    (const guint8 *)&req,
+				    sizeof(req),
+				    (guint8 *)&res,
+				    sizeof(res),
+				    error))
+		return FALSE;
+	if (!fu_igsc_device_heci_validate_response_header(self,
+							  &res.response,
+							  req.header.command_id,
+							  error))
+		return FALSE;
+
+	/* success */
+	*major_vcn = res.major_vcn;
+	*major_version = res.major_version;
+	if (res.oem_version_fitb_valid) {
+		*oem_version = res.oem_version_fitb;
+	} else {
+		*oem_version = res.oem_version_nvm;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_igsc_device_get_subsystem_ids(FuIgscDevice *self, GError **error)
+{
+	struct gsc_fwu_heci_get_subsystem_ids_message_req req = {
+	    .header.command_id = GSC_FWU_HECI_COMMAND_ID_GET_SUBSYSTEM_IDS};
+	struct gsc_fwu_heci_get_subsystem_ids_message_resp res = {0x0};
+	g_autoptr(GError) error_local = NULL;
+
+	if (!fu_igsc_device_command(self,
+				    (const guint8 *)&req,
+				    sizeof(req),
+				    (guint8 *)&res,
+				    sizeof(res),
+				    error))
+		return FALSE;
+
+	/* in older versions this command wasn't supported */
+	if (!fu_igsc_device_heci_validate_response_header(self,
```

Comment:
presumably this is a pre-launch firmware problem right?  Can't this go?

Reply:
waiting for Intel, but I think we'll need to support upgrading from old version to new version as there's quite a bit of review hardware in the field.

Reply:
I've removed this.